tokio/src/net/tcp/stream.rs

```diff
@@ -936,6 +936,168 @@ impl TcpStream {
+    /// ready. Returning a `WouldBlock` error in any other situation will
+    /// incorrectly clear the readiness flag, which can cause the socket to
+    /// behave incorrectly.
+    ///
+    /// The closure should not perform the write operation using any of the
+    /// methods defined on the Tokio `TcpStream` type, as this will mess with
+    /// the readiness flag and can cause the socket to behave incorrectly.
+    ///
+    /// Usually, [`writable()`] or [`ready()`] is used with this function.
+    ///
+    /// [`writable()`]: TcpStream::writable()
+    /// [`ready()`]: TcpStream::ready()
+    pub fn try_write_io<R>(&self, f: impl FnOnce() -> io::Result<R>) -> io::Result<R> {
+        self.io.registration().try_io(Interest::WRITABLE, f)
+    }
+
+    /// Polls for write from the socket using a user-provided IO operation.
+    ///
+    /// If the tcp stream is not currently ready for writing, this method will
+    /// store a clone of the `Waker` from the provided `Context`. When the tcp
+    /// stream becomes ready for writing, `Waker::wake` will be called on the
+    /// waker.
+    ///
+    /// If the tcp stream is ready for writing, the provided closure is called.
+    /// The closure should attempt to write from the socket by manually calling the
+    /// appropriate syscall. If the operation fails because the socket is not
+    /// actually ready, then the closure should return a `WouldBlock` error and
+    /// the read readiness flag is cleared. The `Waker` from the provided `Context`
+    /// will be stored. When the tcp stream becomes ready for writing, `Waker::wake`
+    /// will be called on the waker. The provided closure may be called multiple times
+    /// if the socket become ready right in the instance after the closure has returned
+    /// `WouldBlock`.
+    ///
+    /// Note that on multiple calls to `poll_write_io` only
+    /// the `Waker` from the `Context` passed to the most recent call is
+    /// scheduled to receive a wakeup. (However, `poll_read_io` retains a
+    /// second, independent waker.)
+    ///
+    /// The closure should only return a `WouldBlock` error if it has performed
+    /// an IO operation on the socket that failed due to the socket not being
+    /// ready. Returning a `WouldBlock` error in any other situation will
+    /// incorrectly clear the readiness flag, which can cause the socket to
+    /// behave incorrectly.
+    ///
+    /// The closure should not perform the write operation using any of the
+    /// methods defined on the Tokio `TcpStream` type, as this will mess with
+    /// the readiness flag and can cause the socket to behave incorrectly.
+    ///
+    /// # Return value
+    ///
+    /// The function returns:
+    ///
+    /// * `Poll::Pending` if the tcp stream is not ready for writing.
+    /// * `Poll::Ready(Ok(R))` if the `f` returns `Ok(R)`.
+    /// * `Poll::Ready(Err(e))` if an error is encountered from `f` except `WouldBlock`.
```

Comment:
Copied from `poll_read_io`.
```suggestion
    /// * `Poll::Pending` if the tcp stream is not ready for writing, or if `f` returns a `WouldBlock` error.
    /// * `Poll::Ready(Ok(r))` if `f` returns `Ok(r)`.
    /// * `Poll::Ready(Err(e))` if `f` returns an error other than `WouldBlock`, or if polling for readiness encounters an IO error.
```